### ISSUE TITLE
Reorganise loosely scattered articles in programming guide

### ISSUE THREAD

**Issue by rdb** [panda3d/panda3d-docs#44]
We have a bunch of quite loosely scattered articles in the programming guide section that don't really belong to anything:

- [x] Motion paths
- [x] ShowBase (this information really belongs in the API reference)
- [x] File reading
- [x] Threading
- [x] Subclassing
- [x] Garbage collection, which is in a section all by itself
- [x] Global Clock, a stub that is in a section all by itself, which has a confusing intro
- [x] Matrix Representation, all by itself under Math Engine
- [x] Panda3D Utility Functions (which is about PythonUtil)
- [x] After removing the deprecated patching stuff from the Multifile section, it will contain two pages, one of which is not directly related to multifiles

Most of these pages were clearly added as an after-thought without much thought given to where they fit into the big order of things.  They are also kind of bloating the list and I think some reorganisation would be good to make it easier to find things.  Once we reduce the number of second-level sections, I would like to experiment with laying them out spatially like the [Blender](https://docs.blender.org/manual/en/latest/), [Unity](https://docs.unity3d.com/Manual/index.html) and UE4 manuals do it.

I would suggest removing the ShowBase page, and grouping the stuff about multifiles, asynchronous loading and general file loading in a section that is about "Advanced Loading Techniques" or "Filesystem" or "Pipeline" or something like that, which covers the virtual file system, asynchronous loading, all that stuff.

I think we might want a distinct "async" section covering threading and coroutines, more broken-up and fleshed-out, or we can put this information under the task section, since they are closely related.

I think the incomplete stub about the global clock should go away.  It just lists two methods, and listing methods is the job of the API reference.  Perhaps we could mention the global clock in passing in the Main Loop page under Tasks & Events, since it's relevant to people when writing their tasks and game loops.  We should also mention globalClock and link to ClockObject in the ShowBase API reference.

Motion paths should go under Intervals, since it mainly talks about motion path intervals.  Actually, there is already a page under intervals, which just links to this one, so they should just be merged.

I am not sure what to do with Garbage Collection and Subclassing.  Maybe we could put these together in a section about Python quirks?  Not sure what this would be called.  We could put information about using Python threads and file I/O with Panda3D here too, perhaps.  I'm a little reluctant to have a section about Python quirks because the user is probably unlikely to look for such a section.

Panda3D Utility Functions is by its nature difficult to place.  Though, perhaps it should go under the Panda3D Tools section, since it talks about tools that are only useful interactively during development and not functions that you would ever have in a finished game.

Thoughts?  Other ideas?

**Comment by Moguri** [comment]
Some initial thoughts:

I like the direction this is going in.

For the filesystem stuff, I prefer "Filesytem" or "Advanced Loading" over "Pipeline" since pipeline can be a bit overloaded, and, even when talking about an asset pipeline, I feel like this section might only cover getting data into Panda and not getting data out of other programs.

I think a "Python-specific stuff" category (with a better name) makes sense. Basically, a section on using Panda3D via Python. We may also go into some best practices stuff like virtual environments. However, this starts bleeding into getting started and installation topics.

I don't have any thoughts on utility functions at the moment.

**Comment by rdb** [comment]
I've gone ahead and ticked off many of the items here.  We still need to create the section on "Python quirks" (once I can think of a good thing to call it), find a place for the motion paths stuff, and put some stuff on asynchronous loading under the new Advanced Loading section.

The "Utility Functions" page contained two sections, one about a function that is actually long removed from PythonUtil, and the rest is talking about inspection tools.  I have renamed it to Inspection Utilities and moved it to the Debugging chapter (though we might need to talk about either fleshing that out or merging it with something else, because it is tiny).

**Comment by rdb** [comment]
I renamed the Garbage Collection section to Object Management, under which now also live the pages Subclassing and Reference Counting (the latter was originally from the C++ section, which is also gone now). Not 100% happy this is the best name, but they all relate to OOP / memory management stuff, and it might be a good home for other pages about memory usage and such.

I moved the Threading page to Tasks and Events, it feels like it's conceptually similar enough, but maybe a slight renaming of the Tasks and Events section (scheduling? it shouldn't be too off-putting to beginners) would help it fit better. Or, as suggested earlier, moving all the asynchronous programming pages to a new "asynchronous execution" or "parallelism" section, but I want to be careful not to make async APIs and coroutines feel like a second-class citizen (we rather should encourage people to start with async stuff from the get-go). That said, there is something to be said for keeping "tasks and events" simple for beginners and having the advanced stuff live later on in the manual.

Other ideas for restructuring are welcome, but for now I am closing this issue since I've addressed all the items in the original issue.
